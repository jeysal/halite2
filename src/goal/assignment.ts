import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';
import { maxBy } from 'ramda';

import Goal from '../goal/goal';

interface ShipGoal<Turn> {
  ship: OwnShip<Turn>;
  goal: Goal<Turn>;
}

const VALUE_IMPACT = 0.5;
const SUITABILITY_IMPACT = 0.5;

const shipGoalValue = <Turn>(
  ship: OwnShip<Turn>,
  goal: Goal<Turn>,
  gm: GameMap<Turn>,
): number =>
  VALUE_IMPACT * goal.value(gm) +
  SUITABILITY_IMPACT * goal.suitability(ship, gm);

const assignGoalsToShips = <Turn>(
  goals: Goal<Turn>[],
  gm: GameMap<Turn>,
): ShipGoal<Turn>[] => {
  const ownShips = gm.myShips;

  const shipGoalValues = ownShips.map(ship =>
    goals.map(goal => ({ ship, goal, val: shipGoalValue(ship, goal, gm) })),
  );

  const goalsWithShips = shipGoalValues.reduce((gwss, gVals) => {
    const gValsBelowLimit = gVals.filter(gVal => {
      const entry = gwss.find(gws => gws.goal === gVal.goal);
      return !entry || entry.ships.length < entry.goal.maxShips();
    });
    if (gValsBelowLimit.length) {
      const bestVal = gValsBelowLimit.reduce(maxBy(value => value.val));
      return gwss.map(
        goalWithShips =>
          bestVal && goalWithShips.goal === bestVal.goal
            ? {
                goal: goalWithShips.goal,
                ships: [...goalWithShips.ships, bestVal.ship],
              }
            : goalWithShips,
      );
    }
    return gwss;
  }, goals.map(goal => ({ goal, ships: [] as OwnShip<Turn>[] })));

  // TODO goal minima

  return goalsWithShips.reduce(
    (acc, gws) => [
      ...acc,
      ...gws.ships.map(ship => ({ ship, goal: gws.goal })),
    ],
    [] as ShipGoal<Turn>[],
  );
};

export default assignGoalsToShips;
