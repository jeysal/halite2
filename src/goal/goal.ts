import { Action } from 'halite/Game';
import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';

export default interface Goal<Turn = number> {
  value(gm: GameMap<Turn>): number;
  minShips(): number;
  maxShips(): number;
  suitability(ship: OwnShip<Turn>, gm: GameMap<Turn>): number;
  strive(ship: OwnShip<Turn>, gm: GameMap<Turn>): string & Action<Turn>;
};
