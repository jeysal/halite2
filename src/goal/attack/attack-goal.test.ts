import DockingStatus from 'halite/DockingStatus';
import GameMap from 'halite/GameMap';
import Ship, { EnemyShip } from 'halite/Ship';

import { goalStats } from '../../test/goal';
import { emptyMap } from '../../test/util';

import AttackGoal, { determineAttackGoals } from './attack-goal';

describe('value', () => {
  test('is low for a swarm that just floats around', () => {
    const enemySwarm: EnemyShip[] = [
      new Ship(emptyMap, 1, { dockingStatus: DockingStatus.UNDOCKED }),
      new Ship(emptyMap, 1, { dockingStatus: DockingStatus.UNDOCKED }),
    ] as any;

    expect(new AttackGoal(enemySwarm).value(emptyMap)).toMatchSnapshot();
  });

  test('is high for a swarm that guards a docked planet', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlanets([{ id: 42, x: 15, y: 15 }]);
    const enemySwarm: EnemyShip[] = [
      new Ship(gm, 1, {
        dockingStatus: DockingStatus.DOCKED,
        dockedPlanetId: 42,
      }),
      new Ship(gm, 1, { dockingStatus: DockingStatus.UNDOCKED }),
    ] as any;

    expect(new AttackGoal(enemySwarm).value(gm)).toMatchSnapshot();
  });
});

describe('minShips', () => {
  test('returns a reasonable minimum for a small enemy swarm', () => {
    expect(new AttackGoal(Array(1)).minShips()).toMatchSnapshot();
  });

  test('returns a reasonable minimum for a large enemy swarm', () => {
    expect(new AttackGoal(Array(8)).minShips()).toMatchSnapshot();
  });
});

describe('maxShips', () => {
  test('returns a reasonable maximum for a small enemy swarm', () => {
    expect(new AttackGoal(Array(1)).maxShips()).toMatchSnapshot();
  });

  test('returns a reasonable maximum for a large enemy swarm', () => {
    expect(new AttackGoal(Array(8)).maxShips()).toMatchSnapshot();
  });
});

describe('suitability', () => {
  test('is high for an undocked nearby ship', () => {
    expect(
      new AttackGoal([
        new Ship(emptyMap, 1, { x: 10, y: 10 }),
        new Ship(emptyMap, 1, { x: 12, y: 8 }),
      ] as any).suitability(
        new Ship(emptyMap, 0, {
          x: 20,
          y: 20,
          dockingStatus: DockingStatus.UNDOCKED,
        }) as any,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });

  test('is low for a docked nearby ship', () => {
    expect(
      new AttackGoal([
        new Ship(emptyMap, 1, { x: 10, y: 10 }),
        new Ship(emptyMap, 1, { x: 12, y: 8 }),
      ] as any).suitability(
        new Ship(emptyMap, 0, {
          x: 20,
          y: 20,
          dockingStatus: DockingStatus.DOCKED,
        }) as any,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });

  test('is low for an undocked distant ship', () => {
    expect(
      new AttackGoal([
        new Ship(emptyMap, 1, { x: 10, y: 10 }),
        new Ship(emptyMap, 1, { x: 12, y: 8 }),
      ] as any).suitability(
        new Ship(emptyMap, 0, {
          x: 80,
          y: 80,
          dockingStatus: DockingStatus.UNDOCKED,
        }) as any,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });
});

describe('strive', () => {
  test('navigates towards the enemy swarm', () => {
    expect(
      new AttackGoal([new Ship(emptyMap, 1, { x: 15, y: 15 }) as any]).strive(
        new Ship(emptyMap, 0, { id: 42, x: 50, y: 50 }) as any,
      ),
    ).toMatchSnapshot();
  });

  test('undocks a docked ship', () => {
    expect(
      new AttackGoal([new Ship(emptyMap, 1, { x: 15, y: 15 }) as any]).strive(
        new Ship(emptyMap, 0, {
          id: 42,
          x: 50,
          y: 50,
          dockingStatus: DockingStatus.DOCKED,
        }) as any,
      ),
    ).toMatchSnapshot();
  });
});

describe('determineAttackGoals', () => {
  test('creates an attack goal for each enemy swarm', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlanets([{ id: 1000, x: 73, y: 29, radius: 2, ownerId: 1 }]);
    gm.addPlayerShips(1, [
      {
        id: 100,
        x: 15,
        y: 17,
        radius: 1,
        dockingStatus: DockingStatus.UNDOCKED,
      },
      {
        id: 101,
        x: 13,
        y: 15,
        radius: 1,
        dockingStatus: DockingStatus.UNDOCKED,
      },
      {
        id: 102,
        x: 17,
        y: 16,
        radius: 1,
        dockingStatus: DockingStatus.UNDOCKED,
      },
      {
        id: 103,
        x: 75,
        y: 30,
        radius: 1,
        dockingStatus: DockingStatus.DOCKED,
        dockedPlanetId: 1000,
      },
      {
        id: 104,
        x: 79,
        y: 26,
        radius: 1,
        dockingStatus: DockingStatus.UNDOCKED,
      },
      {
        id: 105,
        x: 65,
        y: 39,
        radius: 1,
        dockingStatus: DockingStatus.UNDOCKED,
      },
    ]);

    expect(
      determineAttackGoals(gm).map(goal =>
        goalStats(goal, new Ship(gm, 0, { id: 42, x: 70, y: 70 }), gm),
      ),
    ).toMatchSnapshot();
  });
});
