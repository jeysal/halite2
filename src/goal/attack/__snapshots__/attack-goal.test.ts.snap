// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`determineAttackGoals creates an attack goal for each enemy swarm 1`] = `
Array [
  Object {
    "action": "t 42 7 226",
    "maxShips": 8,
    "minShips": 4,
    "suitability": 0.6729862388216667,
    "value": 0.41428571428571426,
  },
  Object {
    "action": "t 42 7 279",
    "maxShips": 5,
    "minShips": 1,
    "suitability": 0.819351169392105,
    "value": 0.5086325933483488,
  },
  Object {
    "action": "t 42 7 263",
    "maxShips": 3,
    "minShips": 1,
    "suitability": 0.8667783801329529,
    "value": 0.30000000000000004,
  },
]
`;

exports[`maxShips returns a reasonable maximum for a large enemy swarm 1`] = `20`;

exports[`maxShips returns a reasonable maximum for a small enemy swarm 1`] = `3`;

exports[`minShips returns a reasonable minimum for a large enemy swarm 1`] = `10`;

exports[`minShips returns a reasonable minimum for a small enemy swarm 1`] = `1`;

exports[`strive navigates towards the enemy swarm 1`] = `"t 42 7 225"`;

exports[`strive undocks a docked ship 1`] = `"u 42"`;

exports[`suitability is high for an undocked nearby ship 1`] = `0.9397007462732747`;

exports[`suitability is low for a docked nearby ship 1`] = `0.5397007462732747`;

exports[`suitability is low for an undocked distant ship 1`] = `0.5799571450435087`;

exports[`value is high for a swarm that guards a docked planet 1`] = `0.4891666666666667`;

exports[`value is low for a swarm that just floats around 1`] = `0.3666666666666667`;
